### ISSUE TITLE
Segfault with specific URL but only over HTTPS

### ISSUE THREAD

**Issue by Fossil01** [curl/curl#7932]
### I did this

curl https://webservice.fanart.tv/ -vvvv

### Problem description:
When running this specific URL with curl on this VM running on my M1 Macbook Pro under Parallels it segfaults: `Segmentation fault (core dumped)`. Loading it over http does work though. Also other HTTPS sites seem to work fine, so I am curious why it segfaults at this specfic URL.

Running the same command directly on my Mac or on any other Linux server I have access to works normally.

I am calling this URL with some more parameters from PHP so the problem started there and I narrowed it down to curl used directly on the command line. Not sure if this is curl because wget also segfaults and httpie too but I have no clue where to report it otherwise.

### curl/libcurl version
```
curl 7.68.0 (aarch64-unknown-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```
### operating system

`Linux homestead 5.4.0-89-generic #100-Ubuntu SMP Fri Sep 24 14:29:20 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux`

### GDB
I ran: `gdb --args curl https://webservice.fanart.tv`

Output:
```
Starting program: /usr/bin/curl https://webservice.fanart.tv
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
[New Thread 0xfffff681bf90 (LWP 26383)]
[Thread 0xfffff681bf90 (LWP 26383) exited]

Thread 1 "curl" received signal SIGSEGV, Segmentation fault.
0x0020fffff7a7507c in ?? ()
```

**Comment by bagder** [comment]
You need to rebuild it with debug symbols to get something sensible out of the crash.

**Comment by bagder** [comment]
I built curl 7.68.0 and tested it against this URL and it worked just fine:

curl 7.68.0-DEV (x86_64-pc-linux-gnu) libcurl/7.68.0-DEV OpenSSL/1.1.1l zlib/1.2.11 brotli/1.0.9 c-ares/1.17.0 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.0) nghttp2/1.46.0-DEV librtmp/2.3

**Comment by Fossil01** [comment]
@bagder Could you try it on ARM maybe?

I will see if I can figure out how to build it with debug symbols tonight. But since wget it also affected (and httpie) I guess it's some underlying lib but don't know which.

**Comment by bagder** [comment]
I rather suspect your TLS library install or something is broken. I see no curl bug here.

**Comment by Fossil01** [comment]
Would that be  OpenSSL/1.1.1f  in this case?

**Comment by bagder** [comment]
> Would that be OpenSSL/1.1.1f in this case?

You'd have to do some actual debugging to figure this out for sure. Or of course, if the problem goes away with an upgrade you can probably be reasonably sure it was there.

**Comment by bagder** [comment]
Not able to reproduce, suspected issue in user's environment. Closing.

**Comment by juanparati** [comment]
I also experience the same issue so I statically compiled Curl with OpenSSL/1.1.1f
 instead of 1.1.1l and it works.

**Comment by Fossil01** [comment]
I already was on OpenSSL/1.1.1f which comes with Ubuntu 20.04 LTS for ARM, still have the issue.

**Comment by bagder** [comment]
I think @juanparati wrote that the wrong way around. He reported the crash with OpenSSL/1.1.1f

**Comment by Fossil01** [comment]
Got any pointers @juanparati? Last time I tried it I encountered a fuckton of errors :-)

**Comment by juanparati** [comment]
> I think @juanparati wrote that the wrong way around. He reported the crash with OpenSSL/1.1.1f

Yes I wrote in the wrong way :P  
 edited the comment and now it's correct.

**Comment by juanparati** [comment]
I believe that this is the problem:
https://nvd.nist.gov/vuln/detail/CVE-2020-1967

**Comment by Fossil01** [comment]
Interesting. I wonder why Ubuntu hasn't update the 20.04 repos with a new version.

**Comment by pentatonicfunk** [comment]
had exactly same issue, different URL, but also specific -- https://digitalocean.com. and yes only in ARM

```shell
root@5bf4c10d0a09:/# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.3 LTS"
## arch
root@5bf4c10d0a09:/# arch
aarch64
root@5bf4c10d0a09:/# curl -V
curl 7.68.0 (aarch64-unknown-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets

### DO
root@5bf4c10d0a09:/# curl -I "https://api.digitalocean.com/"
Segmentation fault

### G
root@5bf4c10d0a09:/# curl -I "https://www.google.com/"
HTTP/2 200
content-type: text/html; charset=ISO-8859-1
p3p: CP="This is not a P3P policy! See g.co/p3phelp for more info."
date: Wed, 17 Nov 2021 10:24:51 GMT
server: gws
x-xss-protection: 0
x-frame-options: SAMEORIGIN
expires: Wed, 17 Nov 2021 10:24:51 GMT
cache-control: private
set-cookie: 1P_JAR=2021-11-17-10; expires=Fri, 17-Dec-2021 10:24:51 GMT; path=/; domain=.google.com; Secure
set-cookie: NID=511=RZAfgHD8-XNziica7Vd_cWkDtDMVj1wB6vSkvXUcp0cwSJCIDRDmRgYA38DY7P6g965WVciq52W_WEMLjlASNLjmuXFhkJsV9HItdssoafrQ85Khnb-fRayglF9xIyrfJr9WbgmknmTy9FxpKgITZGYlqsN0QcT-F9EG3CbFN8o; expires=Thu, 19-May-2022 10:24:51 GMT; path=/; domain=.google.com; HttpOnly
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
```

on x86_64, it went fine
```shell
root@de73a2fe7e42:/# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.3 LTS"

## arch
root@de73a2fe7e42:/# arch
x86_64
root@de73a2fe7e42:/# curl -V
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets

#### DO
root@de73a2fe7e42:/# curl -I "https://api.digitalocean.com/"
HTTP/2 301 
date: Wed, 17 Nov 2021 10:24:04 GMT
content-type: text/html
content-length: 102
cache-control: no-cache
location: https://developers.digitalocean.com/
x-gateway: Edge-Gateway
x-request-id: d59a6355-3d51-467e-a065-d94a28e7af28
x-response-from: service
x-runtime: 0.004740
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
set-cookie: __cf_bm=YB1.kK4M8hf.tBv8HJtCxPPY0IkCtsFNDrCdhttNDDQ-1637144644-0-AR2ZJC1HSENOoKOUMjGC97lcrG0aI8QJUXDOUCZsY8uzMNwXCq+Jn7UHqGOYYe3RrWpypi4h9SDvUvNPyod42wmOXRV1rLEvh8vos/gvbbqS; path=/; expires=Wed, 17-Nov-21 10:54:04 GMT; domain=.digitalocean.com; HttpOnly; Secure; SameSite=None
server: cloudflare
cf-ray: 6af83f4e1c976fe9-IAD


### G
root@de73a2fe7e42:/# curl -I "https://www.google.com/"
HTTP/2 200 
content-type: text/html; charset=ISO-8859-1
p3p: CP="This is not a P3P policy! See g.co/p3phelp for more info."
date: Wed, 17 Nov 2021 10:24:24 GMT
server: gws
x-xss-protection: 0
x-frame-options: SAMEORIGIN
expires: Wed, 17 Nov 2021 10:24:24 GMT
cache-control: private
set-cookie: 1P_JAR=2021-11-17-10; expires=Fri, 17-Dec-2021 10:24:24 GMT; path=/; domain=.google.com; Secure
set-cookie: NID=511=UNpxAC39AOLKoGx3srJUTYabxFRNm9a2gFVio_zd07jaxXP4cig4A5hhoTC6ZzaGbKCYnLPkcyHtPgQ0NajmEhitU_4tMnTfmwbJ0wmmUVbw7A1F-AnJ3Ay9aE8qgw5tOjdjlCpJqc_pZFe8urG4ChwDSCvquCe4R-OO_HqNPT0; expires=Thu, 19-May-2022 10:24:24 GMT; path=/; domain=.google.com; HttpOnly
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
```

So i guess its not really Ubuntu @Fossil01 ? because both is Ubuntu 20.04. the diff is only the architecture 

any workaround @juanparati ?
> statically compiled Curl with OpenSSL/1.1.1l instead of 1.1.1f and it works.
how it can be done ? and will it help with php-curl ? or do i need to recompile php-curl too after it ?

**Comment by Fossil01** [comment]
https://digitalocean.com/ segfaults for me too. I tried to compile OpenSSL 1.1.1l but I got errors:

` undefined symbol: EVP_mdc2, version OPENSSL_1_1_0` and I havent been able to fox it yet.

**Comment by juanparati** [comment]
I reported as a bug to Ubuntu Launchpad:
https://bugs.launchpad.net/ubuntu/+source/openssl/+bug/1951279

For those that are facing this problem and need a Homestead environment (@Fossil01: maybe for you) that works with LibCurl and OpenSSL please check:
https://app.vagrantup.com/Juanparati/boxes/homestead-debian-arm

**Comment by juanparati** [comment]
It's confirmed that it's a OpenSSL issue.

I can reproduce this issue executing the following command:

`openssl s_client -showcerts -connect graph.facebook.com:443 </dev/null`

**Comment by davidkhess** [comment]
After spending a lot of time debugging this, I'm pretty confident the root of the problem was fixed in this commit:

https://github.com/openssl/openssl/commit/5795acffd8706e1cb584284ee5bb3a30986d0e75

Note, that bug trashes the stack so that's why it's hard to get a stack trace.

It looks like that fix was landed in OpenSSL 1.1.1i so upgrading to 1.1.1i or newer should fix it.
